formalize how to introduce changes into the compiler (#41)
FuelLabs/sway-rfcs: rfcs/0000-changes-lifecyle.md

The compiler will follow a "rolling release" scheme, which means that periodically (to be specified) a
new major version will be released.

This means that as soon as the compiler reach version "1.0.0", the next **major version** will be "1.1.0"; and
after that and if needed, the next **minor version** would be "1.1.1".

The only changes that will trigger "minor updates" are:
1 - Urgent security fixes;
2 - Bugs that render "stable" functionality unusable;

## PR changes

Every PR will have a section named "Release Notes", which will contains a user friendly explanation of what the PR contains.
That allows the whole team to understand the impact of the change, if it contains breaking changes or not, and correctly
configure the PR to the next major or minor update.

The compiler release notes **WILL** be the amalgamation of these section "as is".

## Breaking changes

A breaking change is a difference in functionality from the previous version of the compiler that may require an update to sway code in order for it to compile.

The following changes are defined to be breaking changes, and will need to follow the process of being gated by features flags.

> 1 - Code without deprecated warning to be flagged as error;
> 2 - ABI json properties to be removed or have their type changed;
> 3 - Binary encoding that will break how SDK (and others) communicate with the compiler, that includes
> 

Reviewer: ```suggestion
- Code without deprecated warning to be flagged as error;
- ABI json properties to be removed or have their type changed;
- Binary encoding that will break how SDK (and others) communicate with the compiler, that includes
```

formatting nits

- How scripts/predicates accept arguments on `main`;
- How scripts/predicates return data from `main`;
- How the `contract method selector` is encoded;
- How `contract method` arguments are encoded;
- How `log` data is encoded;
- How `message` data is encoded;

>   4 - Utilization of new VM opcodes;
>   5 - IR changes
>   6 - Receipt parsers to break;
>   7 - When a compiler feature or a standard library produce different behavior for the same code (semantic changes);

Reviewer: ```suggestion
- Utilization of new VM opcodes;
- IR changes
- Receipt parsers to break;
- When a compiler feature or a standard library produce different behavior for the same code (semantic changes);
```

formatting nits

> 1 - Code without deprecated warning to be flagged as error;
> 2 - ABI json properties to be removed or have their type changed;
> 3 - Binary encoding that will break how SDK (and others) communicate with the compiler, that includes
> 
> - How scripts/predicates accept arguments on `main`;
> - How scripts/predicates return data from `main`;
> - How the `contract method selector` is encoded;
> - How `contract method` arguments are encoded;
> - How `log` data is encoded;
> - How `message` data is encoded;
> 
>   4 - Utilization of new VM opcodes;
>   5 - IR changes
>   6 - Receipt parsers to break;
>   7 - When a compiler feature or a standard library produce different behavior for the same code (semantic changes);

Reviewer: The formatting seems really off here:

```suggestion
1. Code without deprecated warning to be flagged as error;
1. ABI json properties to be removed or have their type changed;
1. Binary encoding that will break how SDK (and others) communicate with the compiler, that includes
   - How scripts/predicates accept arguments on `main`;
   - How scripts/predicates return data from `main`;
   - How the `contract method selector` is encoded;
   - How `contract method` arguments are encoded;
   - How `log` data is encoded;
   - How `message` data is encoded;
1. Utilization of new VM opcodes;
1. IR changes
1. Receipt parsers to break;
1. When a compiler feature or a standard library produce different behavior for the same code (semantic changes);
```

Author: Done.

- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [FuelLabs/sway-rfcs#0000](https://github.com/FuelLabs/sway-rfcs/pull/001)
- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)

# Summary

[summary]: #summary

This RFC intends to propose a scheme to track changes in the compiler. 

# Motivation

[motivation]: #motivation

The compiler needs a guide on how to introduce changes to improve developer experience every time a compiler is updated.

# Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

## PR changes

> Every PR will have a section named "Release Notes", which will contains a user friendly explanation of what the PR contains.

Reviewer 2: We should probably introduce the rest of the changeset benefits, including specifying the level of change the PR does (minor, major, patch, etc).

Reviewer 3: Have we considered putting the release notes into a changelog file, rather than having them spread across different PR descriptions?

Author: For the moment I moved this to the "open questions" section.

Reviewer 3: Another option is to generate the changelog file from commits as part of the release process, using something like https://git-cliff.org/

We'd have to be more rigorous about commit messages, i.e. using `feat!:` to indicate a breaking change for a feature.

The compiler release notes **WILL** be the amalgamation of these section "as is".

> 

Reviewer 2: I also think that we should discuss formalizing what our policy towards breaking changes is here.

Some of this is getting discussed elsewhere for other moving parts of the ecosystem, but I think we should commit to not making breaking changes outside of specific timed releases and emergencies (security related bugs, etc).

The normal process should be that a breaking changes goes through behind a feature flag and those features get stabilized at predictable intervals when we do a breaking release.

5 - On a later date, the feature flag can be removed making the feature the default behavior of the compiler. 

# Enabling features on `sway`

Features can be enabled inside the `Forc.toml` file:

```toml
experimental = ["...", "..."]
```

or using the CLI

```
> forc ... --experimental some_feature,another_feature
```

These flags also need to be enabled programmatically by any compiler driver, like tests.

Unlike `Rust` we will not support features inside sway code like the example below, because some features will span across multiple tools. That would demand `forc` to parse, or ask the `sway` compiler if a feature is enabled or not. 

```sway
#![enable(some_experimental_feature)]
```

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

This RFC intends to propose a scheme to track changes in the compiler. These changes will be categorized into two ways:

1 - Non-breaking changes;
2 - Breaking changes.

Breaking changes will be any changes that, after updating the compiler version, causes:

1 - Code without deprecated warning to be flagged as error;
2 - ABI json properties to be removed or have their type changed;

> 3 - SDK (rust, typescript and others) to break;

Reviewer 2: This includes it implicitly but we should explicitly mention noncompatible changes to the ABI (not the spec, the actual binary interface of the contracts) and VM incompatibility.

4 - Receipt parsers to break;

The following changes are **not** considered to be breaking changes:

1 - Code flagged with deprecated warning to be flagged as error;

> 2 - Code that required an experimental flag, but now are the default behavior to compile;

Reviewer 2: This I don't agree with, stabilizing an experimental flag should be a breaking change, if it is indeed noncompatible.

> 3 - Contract ID to change;

Reviewer 3: This is also listed under Breaking Changes 😆

4 - new ABI json properties;

> 

Reviewer 2: This should also include IR changes. (at least for now)

## Minor Version Updates

> When a developer updates to a different minor version, it is expected no breaking changes, except for severe bugs.

Reviewer 2: ```suggestion
When a developer updates to a different minor version, no breaking changes are expected, except for severe bugs.
```

Minor changes will consider these to be breaking and non-breaking changes:

> ### Breaking Changes 
> 

Reviewer 3: It might be good to start off this section with a definition of breaking changes, before listing the categories of breaking changes. Agreeing on a definition will make it easier to debate the categorization.

Here's a definition to start with.

```suggestion
### Breaking Changes 

A breaking change is a difference in functionality from the previous version of the compiler that may require an update to sway code in order for it to compile.
```

> 1 - If any new warning, error or other diagnostic are emitted;

Reviewer 2: This I don't agree with. Adding new warnings and errors should not necessarily be a breaking changes.

New warnings should always be a nonbreaking change.

New errors should only be a breaking change if previously valid code is now considered erroneous. Invalid sway code that produces new or different errors is not a breaking change.

Reviewer 3: New errors could require sway devs to "migrate" to the new release of the compiler, which I think is the definition of a breaking change. New warnings are not necessarily breaking, but for projects using `--error-on-warnings`, they are essentially the same as new errors.

Author: My though is aligned with what @sdankel said. 

I expect every team to treat warnings as errors, and in these cases, new warnings will force developers to change code.

For the moment I moved this to the "open questions" section.

2 - If any ABI json properties are removed or have their type changed;
3 - If any code using SDK (rust, typescript and others) stop working;
4 - If any receipt cannot be parsed anymore;

> 5 - If any "contract id" changes;

Reviewer 2: This we should discuss. I don't think a contract compiling to a new contract id should necessarily be considered a breaking change. There is an argument for it but I don't see it as breaking because the code still works and compiles as intended.

If you want reproducible builds, you should be using the same exact version of the compiler anyways. So I don't see too much of a benefit in making this breaking.

Reviewer 3: I agree with this, but am curious what the SwayEx devs would say. It would break some of our tests, which means it possibly breaks other sway dev's integration tests.

Author: Moving this to the open questions section.

> 6 - Semantic changes

Reviewer 2: This is too vague. Should be explicited as producing different runtime behavior for the same code. Excluding bugs.

### Non-breaking Changes

1 - new ABI json properties;

> 2 - Changing one warning by another, or one error by another;

Reviewer 3: I think this is clearer:

```suggestion
2 - Changing the name or contents of an error or warning;
```

3 - Code that was **NOT** compiling, but now is.

## Major Version Updates

Major update is the opportunity to introduce breaking changes, but they cannot be introduced abruptly, and depending on the change, will need to be introduced into phases.

Changes that only affect the compiler/language, can be introduced freely following sway pace. Changes that interact with others tools, will need to be introduced into phases and in coordination with others tools.

> These changes will need to be behind "feature flags", which will be described later.

Reviewer 3: ```suggestion
These changes will need to be behind "feature flags", which are described above.
```

# Rationale and alternatives

[rationale-and-alternatives]: #rationale-and-alternatives

There only two other alternatives, which we consider worst options:

1. Keep experimental features in branches;
2. Keep experimental features behind compiler flags (conditional compilation); 

The first one normally creates more problems than it solves. Integration hell become a reality sooner than later.
The second does not decrease complexity of the code, and it decreases testability of features, given that they will not be in the "release binary".

# Prior art

[prior-art]: #prior-art

These are the sources and stacks used as motivation for this RFC.

## How `rustc` does it

https://rustc-dev-guide.rust-lang.org/implementing_new_features.html#stability-in-code

1. - Open a tracking issue;
2. - Pick a name for the feature gate;
3. - Add the feature gate to the compiler code;
4. - If the feature gate is not set, you should either maintain the pre-feature behavior or raise an error, depending on what makes sense;
5. - Add a test to ensure the feature cannot be used without a feature gate;

> 6. - Add a section to the unstable book;

Reviewer 4: Do we also want to have _The Sway Unstable Book_? IMO, yes. It will be easier to communicate to devs what are we working on, but even more important, for us, we can craft the feature documentation as we go and save us from stress of documenting everything on the end of the development.

When a version is being launched, for example `1.10` a new section `1.11-nightly` will be created empty. From this time on, all new PRs will create a new item under the "next version", `1.11` in this example.

Each item will follow the template:

```
# Version 1.11-nightly

## Forc

### Breaking Changes
### New Features
### Bugs Fix
### Others

## Tools

## Sway

## `sway-lib-core` and `sway-lib-std`

# Version 1.10

...
```

Inside each section, items will have the following templae

```
- Index operator using Index trait [#6356](https://github.com/FuelLabs/sway/pull/6356)
```

Given that all PRs will touch this file, to avoid conflicts and decrease the experience when merging them, we will use "git custom merge driver" (see https://git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver).

This allows a custom merge strategy to a specific file using ".gitattributes".

```
ReleaseNotes.md merge=union
```

## Breaking changes

A breaking change is a difference in functionality from the previous version of the compiler that may require an update to sway code in order for it to compile.

> The following changes are defined to be breaking changes, and will need to follow the process of being gated by features flags.

Reviewer 4: Perhaps additional one: changing the way how `storage` field slots are calculated? Right now, we calculate them as `sha256("storage::ns1::ns2::..::nsN.field_name")` and want to introduce a new algorithm: `sha256((0u8, "storage::ns1::ns2::..::nsN.field_name"))`. There were questions on how slots/storage keys are calculated for `storage` fields, so I assume there are devs who rely on that.

- Feature Name: changes-lifecycle
- Start Date: 2024-08-05
- RFC PR: [FuelLabs/sway-rfcs#41](https://github.com/FuelLabs/sway-rfcs/pull/41)

# Summary

[summary]: #summary

This RFC is a guide on how to introduce breaking changes following the compiler release lifecycle.

# Motivation

[motivation]: #motivation

To maximize user experience when updating the compiler, we need a guide on how to introduce breaking changes.

# Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

The compiler will follow a "rolling release" scheme, which means that periodically (to be specified) a
new major version will be released.

> This means that as soon as the compiler reaches version "1.0.0", the next **major version** will be "1.1.0"; and

Author: Hum...  I think `major.minor.patch` makes more sense. Not sure what would be a major increase in our case.

> This means that as soon as the compiler reaches version "1.0.0", the next **major version** will be "1.1.0"; and

Author: I changed to a more neutral wording, just calling for "next version" and "next patch".

> This means that as soon as the compiler reaches version "1.0.0", the next **major version** will be "1.1.0"; and

Reviewer 5: This section is possibly the most important part of the document as it deals with how we actually handle the versioning of breaking changes, but I think I'm still a little confused. I'm hoping the following example can help me understand:

Let's say we're currently live at version `X.Y.Z`, and we have a new release ready to go containing the following in its release notes:

```
### Breaking Changes

- This release removes the ABC feature
```

Seeing as the next release will be a breaking change, what will be the SemVer for its release? Will it be:

- (X+1).Y.Z
- X.(Y+1).Z
- X.Y.(Z+1)

In SemVer this release will be `(X+1).Y.Z` but as I read this current section the rolling release will put it at `X.(Y+1).Z`? Whatever the version we use is not my point here - I'm just thinking we need to be more explicit here with whichever way we decide so that it's not ambiguous in the future.

To be more explicit, maybe a new sub-section should be added here called "Semantic Versioning" which summarises how we will categorise/number each type of change i.e something like:

### Semantic Versioning

The compiler will follow a "rolling release" scheme, meaning that each new release (happening periodically but yet to be specified) will steamroll over all previous releases.

#### MAJOR

`MAJOR` number changes are strictly reserved to deal with dramatic differences in the Sway Language.

#### MINOR

Every normal change, which includes **_all_** breaking changes, new features, and bug fixes, will be a bump increase in the `MINOR` number.

#### PATCH

`PATCH` number changes are out-of-band, and are strictly reserved for security fixes to each currently supported `MAJOR.MINOR` release. In other words, if a security fix is needed, every existing supported release will include a new patch bump release.
```

> This means that as soon as the compiler reaches version "1.0.0", the next **major version** will be "1.1.0"; and

Reviewer 4: I like @alfiedotwtf's detailed wording on major, minor, and patch. Only the title "Semantic Versioning" I find potentially confusing because, at least in my head, it immediately associates on semver where the expectation is to increase the major on breaking changes while in our case, the major version is indeed more a "marketing version". Perhaps just title it "Versioning Scheme"?

after that and if needed, the next **minor version** would be "1.1.1".

> The only changes that will trigger "minor updates" are:

Author: Done.

> 1 - Urgent security fixes;

Author: Done.

> 2 - Bugs that render "stable" functionality unusable;

Author: Done.

## Release Notes

> To track all "release notes" it will be created a special file `ReleaseNotes.md` that will live in the repo root.

Reviewer 6: ```suggestion
To track all release notes, a special file named `ReleaseNotes.md` will be created in the repository's root directory.
```

Author: Done.

> When a version is being launched, for example, `1.10` a new section `1.11-nightly` will be created empty. From this time on, all new PRs will create a new item under the "next version", `1.11` in this example.

Author: Done.

> Each item will follow the template:

Author: The idea was a user friendly one line description. I have changed the template description below.

> Each item will follow the template:

Author: Done.

```
# Version 1.11-nightly

## Forc

### Breaking Changes
### New Features
### Bugs Fix

> ### Others

Author: Done.

## Tools

## Sway

> ## `sway-lib-core` and `sway-lib-std`

Author: Done.

2. A preliminary PR enabling the feature flag should be created and linked to the umbrella issue; This PR will enable the feature flag `--experimental <comma-separated-list>` on all tools.
3. As many PRs will be created and merged as normal;
4. A chapter inside `Sway Unstable book` will be created and updated as needed;
5. When the feature is ready, a closing PR will be created and wait until the feature flag is enabled by default.
6. On a later date, the feature flag can be removed making the feature the default behavior of the compiler.

Once the feature is merged into `master`, it will not be possible to "turn off" the feature. In the same sense
that it is not possible to "turn off" a "match expression" or any other language feature.

If the feature contains some configuration or choice, the "default" value will be the default, and other options
will be available, but it cannot be turned off.

## Conditional compilation

There are cases where conditional compilation will be needed. For these cases, each experimental feature will also have a corresponding `#[cfg(...)]`, like the example below:

> ````sway

Author: Done.

#[cfg(experimental_new_encoding = false)]
const CONTRACT_ID = 0x14ed3cd06c2947248f69d54bfa681fe40d26267be84df7e19e253622b7921bbe;
#[cfg(experimental_new_encoding = true)]
const CONTRACT_ID = 0x316c03d37b53eaeffe22c2d2df50d675e2b2ee07bd8b73f852e686129aeba462;

> ``

Author: Done.

> # Enabling features on `sway`

Author: Done.

Features can be enabled inside the `Forc.toml` file:

```toml
experimental = ["...", "..."]
````

or using the CLI

```
> forc ... --experimental some_feature,another_feature
```

> or event using the environment variables

Author: Done.

> or event using the environment variables

Author: `env vars` are normally more useful when one does not have access to the CLI. FOr example: CI, or when one is compiling a project using a SDK that do not allow `forc` to be configured. Then `env vars` become more useful.

> or event using the environment variables

Author: > Do we fully override or aggregate?

```
> FORC_EXPERIMENTAL_SOME_FEATURE forc ...
```

These flags also need to be enabled programmatically by any compiler driver, like tests.

> Unlike `Rust` we will not support features inside sway code like the example below, because some features will span across multiple tools. That would demand `forc` to parse, or ask the `sway` compiler if a feature is enabled or not.

Author: Done.

If a feature requires or allows new configurations, these configurations should be optional, and as soon as confirmed the feature is enabled, it should verify if the required configuration is available.

```
> forc ... --experimental a_feature --a-feature-option 1
```

## Lexer and Parser

To allow as user-friendly error messages as possible, in all possible cases, we want both the lexer and the parser to parse the new syntax even with the feature off.

After that, the lexer and the parser will mark that an experimental feature was lexed/parsed; and a check will guarantee that no disabled experimental was parsed.

The error message will have a message explaining that the feature is experimental and a Github link for more details on the stabilization lifecycle.

```rust
// always parse new syntax
let new_syntax = parse_new_syntax();
if ctx.experimental.is_disabled(Features::NEW_FEATURE) {
   handler.emit_error(...);
}
```

## Formatting

Formatting should always format new syntax, even when the flag is off. To avoid breaking other parts of the code, or even worse, removing code.

## LSP

> LSP can take advantage of specific error messages, and suggest user to enable the corresponding feature.

Author: Actually went with "suggest user". What do you think?

## CST, AST and Typed Tree

All trees must always support new features, which means that new nodes will always exist. Their specific behavior, desugaring, etc... will be gated by the experimental feature.

More specifically, this means that variants should not be behind compiler flags.

> ## Tests

Reviewer 4: On the CI, are we going to run tests for all the features? What about the combinations of the features? We don't need to define that in this RFC, perhaps it's too early. But it will be an important practical question. One possibility would be to have a CI run with all experimental features off, and one with all on.

Author: What is going to define the configuration is the `test.*.toml`. The default `test.toml` will run the compiler with all experimental off (the same as no flags).

When a specific test needs a specific feature it will have a `test.feature_a.toml`, where that feature will be enabled. Now running our tests will run this specific test twice.

If we force a specific feature we can use the environment variable:

```
> FORC_EXPERIMENTAL=feature_a cargo r -p test --release
```

9. - Add a section to the unstable book;

## changeset

https://github.com/changesets/changesets

# Unresolved questions

[unresolved-questions]: #unresolved-questions

## Where should we save "release notes"?

1. In files inside the repo? - Saving in files sounds like the best approach, but if we use the same file, all PRs will conflict. To avoid this we can do like `changeset` and use random names.
2. In GitHub PR descriptions? - This is the easiest approach, but it can be cumbersome to recover these messages.
3. In git commit messages? - Given that the commit message is "created" when the PR is merged, there is no way to guarantee that "release notes" will exist or be "parseable" when we need them.
   - https://git-cliff.org/

## Should new warnings be considered breaking changes?

1. Normally warnings should not be considered breaking changes, because they do not break anything.
2. on the other hand, if the team treats warnings as errors, new warnings will break CIs, and demand
   developers' attention. Not all fixes are trivial and can demand bigger code changes than the user would expect
   from a minor update from the compiler.

## How to deal with "contract id"?

> Should we consider contract ID changes to be breaking changes?

Author: That is also the general feeling at the moment. 

But I think that there is a chance users will bring two complaints:
1 - update manually is annoying and error-prone;
2 - if the "contract id" changed because of a change I don´t "care/want", I would need to deploy the contract again incurring costs. Or have a complex table like this contract must be compiled with this version, and that contract with that version etc...

We are also considering a tool to help update contract IDs where needed.
